components/camel-kafka/src/main/java/org/apache/camel/processor/idempotent/kafka/KafkaIdempotentRepository.java

```diff
@@ -0,0 +1,342 @@
+
+        Properties producerConfig = new Properties();
+        producerConfig.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+
+        this.topic = topic;
+        this.consumerConfig = consumerConfig;
+        this.producerConfig = producerConfig;
+        this.cache = Collections.synchronizedMap(new LRUCache<>(maxCacheSize));
+    }
+
+    public KafkaIdempotentRepository(String topic, Properties consumerConfig, Properties producerConfig) {
+        this(topic, consumerConfig, producerConfig, DEFAULT_MAXIMUM_CACHE_SIZE);
+    }
+
+    public KafkaIdempotentRepository(String topic, Properties consumerConfig, Properties producerConfig, int maxCacheSize) {
+        StringHelper.notEmpty(topic, "topic");
+        this.topic = topic;
+        ObjectHelper.notNull(consumerConfig, "consumerConfig");
+        this.consumerConfig = consumerConfig;
+        ObjectHelper.notNull(producerConfig, "producerConfig");
+        this.producerConfig = producerConfig;
+        this.cache = Collections.synchronizedMap(new LRUCache<>(maxCacheSize));
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        log.info("Context: {}", camelContext);
+
+        // each consumer instance must have control over its own offset, so assign a groupID at random
+        String groupId = UUID.randomUUID().toString();
+        log.debug("Creating consumer with {}[{}]", ConsumerConfig.GROUP_ID_CONFIG, groupId);
+
+        consumerConfig.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+        consumerConfig.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, Boolean.TRUE.toString());
+        consumerConfig.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        consumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+
+        consumer = new KafkaConsumer<>(consumerConfig);
+
+        producerConfig.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        producerConfig.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        // set up the producer to remove all batching on send
+        producerConfig.putIfAbsent(ProducerConfig.ACKS_CONFIG, "1");
+        producerConfig.putIfAbsent(ProducerConfig.BATCH_SIZE_CONFIG, "0");
+        producer = new KafkaProducer<>(producerConfig);
+
+        cacheReadyLatch = new CountDownLatch(1);
+        topicPoller = new TopicPoller(consumer, cacheReadyLatch);
+    }
+
+    @Override
+    public void setCamelContext(CamelContext camelContext) {
+        // doStart() has already been called at this point
+        this.camelContext = camelContext;
+        ExecutorServiceManager executorServiceManager = camelContext.getExecutorServiceManager();
+        executorService = executorServiceManager.newFixedThreadPool(this, "KafkaIdempotentRepository", 1);
+        executorService.submit(topicPoller);
```

Comment:
The cache, latched at line 280, is warmed up by the poller thread, so this is in the right place.